use-shopping-cart/src/reducers.js

```diff
@@ -139,6 +142,15 @@ export function cartValuesReducer(state, action) {
     case 'clear-cart':
       return cartValuesInitialState
 
+    case 'load-cart':
+      if (!action.shouldMerge) state = { ...cartValuesInitialState }
+
+      for (const sku in action.cartDetails) {
+        const entry = action.cartDetails[sku]
```

Comment:
@dayhaysoos This should provide the ability to filter out cart items that are no longer needed or valid.

```suggestion
        const entry = action.cartDetails[sku]
        if (action.filter && !action.filter(entry))
          continue

```

Reply:
All we need is tests for this if this looks good to you.

Reply:
Could you walk me through how we'd apply these filters? I can't quite grasp what it would look like using this API to filter out invalid cart items

Reply:
@dayhaysoos Once we've modified the `loadCart` method's definition to support a `filter` we could then just say something like:

```js
const { loadCart } = useShoppingCart()

// somehow fetch stored cart data
loadCart(cartDetails, { filter: ifInStock, shouldMerge: true })
```

Where `ifInStock` is just a function that checks to see if a product is in stock like:

```js
async function ifInStock({ sku }) {
  const data = await fetch(baseURL+'/api/checkStock?sku=${sku}').then(res => res.json())
  return data.isInStock
}
```

Note: I think we'll need to move the handling of filtering out of the reducer and into `loadCart()` itself to support asynchronous filtering. Or, we could allow the developer to filter their `cartDetails` themselves by providing a `filter()` function.

```js
export function filterCart(cartDetails, filter) {
  const filteredCart = {}
  for (const sku in cartDetails) {
    const entry = cartDetails[sku]
    if (filter(entry))
      filteredCart[sku] = entry
  }
}
```

Reply:
I think the idea about creating a `filterCart()` function is probably the best idea since it means we don't have to worry about complicating the rest of useShoppingCart and it's super easy to make asynchronous.